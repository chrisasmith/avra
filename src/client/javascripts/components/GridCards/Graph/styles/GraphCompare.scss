@import '../../../../../stylesheets/modules/theme.scss';

.container {
  display: flex;
  position: relative;
  margin-bottom: 14px;

  &:last-child {
    margin-bottom: 0;
  }
}

.graphWrapper {
  width: calc(100% - 70px);
}

.title {
  width: 116px;
  font-size: 15px;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  font-size: 14px;
  line-height: 14px;
  padding-right: 4px;
}

.graph {
  margin-left: 116px;
  flex-grow: 1;
}

.graphLine {
  background-color: #d3dcf6;
  height: 15px;
  float: left;

  &.first {
    background-color: #5071dc;
  }
}

.graphAmount {
  float: left;
  font-size: 14px;
  line-height: 15x;
  vertical-align: top;
  margin-left: 8px;

  &.first {
    font-weight: bold;
  }
}

.lineContainer {
  height: 15px;
  line-height: 15px;
}
