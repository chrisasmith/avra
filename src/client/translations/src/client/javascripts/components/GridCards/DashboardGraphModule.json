[
  {
    "id": "dashboard_graph.total",
    "defaultMessage": "Total"
  },
  {
    "id": "dashboard_graph.see_more",
    "defaultMessage": "See More Results"
  },
  {
    "id": "dashboard_graph.see_less",
    "defaultMessage": "See Less Results"
  }
]