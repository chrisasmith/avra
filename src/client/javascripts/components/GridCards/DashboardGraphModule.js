import React, { Component } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import domToImage from 'dom-to-image'
import format from 'string-format'
import { FormattedMessage } from 'react-intl'
import styles from './styles/DashboardModule.scss'
import GraphCompare from './Graph/GraphCompare'
import ChartGraph from './Graph/ChartGraph'
import { getNumberLevel, formatNumberByLevel } from '../../services/utils'
import DashboardShellModule from './DashboardShellModule'

export const MODULE_TYPE = {
  PERCENT: 'PERCENT',
  COUNT: 'COUNT',
  STRING: 'STRING',
}

const MAX_NUMBER_OF_VALUES = 4

const propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  isFetching: PropTypes.bool.isRequired,
  type: PropTypes.string.isRequired,
  propertyTitle: PropTypes.string,
  benchmarkTitle: PropTypes.string,
  sideTitle: PropTypes.string,
  chart: PropTypes.bool,
  showTotal: PropTypes.bool,
  data: PropTypes.object,
}

class DashboardGraphModule extends Component {
  constructor(props, context) {
    super(props, context)

    this.graph = React.createRef()

    this.onDownloadImage = this.onDownloadImage.bind(this)
    this.onDownloadCSV = this.onDownloadCSV.bind(this)
    this.onExpandClick = this.onExpandClick.bind(this)

    this.state = {
      expanded: false,
    }
  }

  onDownloadImage() {
    const { title } = this.props

    domToImage.toJpeg(this.graph.current, { quality: 0.95 })
      .then((dataUrl) => {
        const link = document.createElement('a')
        link.download = `${title}.jpeg`
        link.href = dataUrl
        link.click()
      })
  }

  onDownloadCSV() {
    const { title, data, propertyTitle, benchmarkTitle } = this.props

    const csvContent = data.values.reduce((p, c) => {
      const row = c.values.join(',')
      p += `${c.title},${row},\r\n`
      return p
    }, `data:text/csv;charset=utf-8,,${propertyTitle},${benchmarkTitle},\r\n`)

    const encodedUri = encodeURI(csvContent)
    const link = document.createElement('a')
    link.setAttribute('href', encodedUri)
    link.setAttribute('download', `${title}.csv`)
    document.body.appendChild(link)

    link.click()
  }

  onExpandClick(e) {
    if (e) e.preventDefault()

    this.setState({ expanded: !this.state.expanded })
  }

  render() {
    const {
      title,
      chart,
      sideTitle,
      description,
      isFetching,
      type,
      propertyTitle,
      benchmarkTitle,
      showTotal,
      data,
    } = this.props


    if (!data || isFetching) {
      return (
        <DashboardShellModule isFetching title={title} />
      )
    }

    const { expanded } = this.state

    const values = data.values.sort((a, b) => b.values[0] - a.values[0])
    const keyDescription = data.key
    const total = data.total

    let max = values.reduce((p, c) => {
      const valueMax = c.values.reduce((pp, cc) => pp > cc ? pp : cc, 0)
      return p > valueMax ? p : valueMax
    }, 0)

    // find our levels
    const level = getNumberLevel(max)
    max = formatNumberByLevel(max, level)

    const decimals = type === MODULE_TYPE.PERCENT ? 0 : 2

    // parse all the values using the levels
    const parsedValues = values.filter((v, i) => expanded ? true : i < MAX_NUMBER_OF_VALUES).map(val => (
      {
        ...val,
        values: val.values.map(v => formatNumberByLevel(v, level, decimals).toFixed(decimals)),
      }
    ))

    // parse total
    const totalLevel = getNumberLevel(total || 0)

    return (
      <DashboardShellModule
        isFetching={isFetching}
        title={title}
        onDownloadCSV={this.onDownloadCSV}
        onDownloadImage={this.onDownloadImage}
      >
        <div>
          {showTotal && (
            <div className={classnames('p-3 text-center ', styles.totalContainer)} ref={this.graph}>
              <div className="pb-3">
                <FormattedMessage
                  id="dashboard_graph.total"
                  defaultMessage="Total"
                />
              </div>
              <h1>{formatNumberByLevel(total || 0, totalLevel, 1)}</h1>
              <div><em>{totalLevel.single}</em></div>
            </div>
          )}
          <div className={classnames(styles.graphWrapper)}>
            <div className={classnames('p-3', styles.graphContainer)} ref={this.graph}>
              {description && (
                <div className="text-center pb-3">
                  {description}
                </div>
              )}
              <div className={styles.graphHolder}>
                {sideTitle && (
                  <div className={styles.rotatedTitle}>
                    <div className={styles.centered}>
                      <em>{sideTitle}</em>
                    </div>
                  </div>
                )}
                <div>
                  {chart ? (
                    <ChartGraph
                      percent={type === MODULE_TYPE.PERCENT}
                      source={propertyTitle}
                      labels={parsedValues.map(v => v.title)}
                      values={parsedValues.reduce((p, c) => [...p, c.values[0]], [])}
                      compareValues={parsedValues.reduce((p, c) => [...p, c.values[1]], [])}
                    />
                  ) : (
                    <div>
                      {parsedValues.map((graph, idx) => (
                        <GraphCompare
                          key={idx}
                          title={graph.title}
                          values={graph.values}
                          type={type}
                          maxValue={max}
                        />
                      ))}
                    </div>
                  )}
                </div>
              </div>
              {keyDescription && (
                <div className="p-3 text-center">
                  <em>{format(keyDescription, level.plural)}</em>
                </div>
              )}
              {propertyTitle && (
                <div className={styles.keys}>
                  <div>
                    <div className={styles.valueKey} />
                    <div>{propertyTitle}</div>
                  </div>
                  <div>
                    <div className={styles.benchKey} />
                    <div>{benchmarkTitle || 'Benchmark'}</div>
                  </div>
                </div>
              )}
            </div>
            {values.length > MAX_NUMBER_OF_VALUES && (
              <div className={styles.seeMoreLink}>
                <a href="#" onClick={this.onExpandClick}>{!expanded
                  ?
                  <FormattedMessage
                    id="dashboard_graph.see_more"
                    defaultMessage="See More Results"
                  />
                  : <FormattedMessage
                    id="dashboard_graph.see_less"
                    defaultMessage="See Less Results"
                  />
                }</a>
              </div>
            )}
          </div>
        </div>
      </DashboardShellModule>
    )
  }
}

DashboardGraphModule.propTypes = propTypes
export default DashboardGraphModule
