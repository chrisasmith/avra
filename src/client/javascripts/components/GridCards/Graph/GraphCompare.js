import React from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import styles from './styles/GraphCompare.scss'
import { MODULE_TYPE } from '../DashboardGraphModule'

const propTypes = {
  title: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
  type: PropTypes.string.isRequired,
  maxValue: PropTypes.number,
}

const GraphCompare = ({ title, values, type, maxValue }) => (
  <div className={styles.container}>
    <div className={styles.title}>{title}</div>
    <div className={styles.graph}>
      {values.map((value, idx) => {
        const percent = type === MODULE_TYPE.PERCENT || !maxValue ? value : (value * 100) / maxValue
        return (
          <div className={styles.lineContainer} key={idx}>
            <div>
              <div className={styles.graphWrapper}>
                <div
                  style={{ width: `${percent}%` }}
                  className={classnames(styles.graphLine, { [styles.first]: idx === 0 })}
                />
              </div>
              <div
                className={classnames(styles.graphAmount, { [styles.first]: idx === 0 })}
              >
                {value}{type === MODULE_TYPE.PERCENT ? '%' : ''}
              </div>
            </div>
          </div>
        )
      })}
    </div>
  </div>
)

GraphCompare.propTypes = propTypes
export default GraphCompare
