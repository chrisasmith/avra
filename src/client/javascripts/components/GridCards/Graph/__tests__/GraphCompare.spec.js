import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import GraphCompare from '../GraphCompare'
import { MODULE_TYPE } from '../../DashboardGraphModule'

describe('GraphCompare Component', () => {
  it('Should show', () => {
    const wrapper = shallow(<GraphCompare
      title="title"
      values={['8', '2']}
      type={MODULE_TYPE.COUNT}
      maxValue={10}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show percent', () => {
    const wrapper = shallow(<GraphCompare
      title="title"
      values={['80', '20']}
      type={MODULE_TYPE.PERCENT}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show edge cases', () => {
    const wrapper = shallow(<GraphCompare
      title="title"
      values={['0', '100']}
      type={MODULE_TYPE.PERCENT}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
