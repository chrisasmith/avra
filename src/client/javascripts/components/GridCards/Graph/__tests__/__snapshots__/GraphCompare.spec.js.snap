// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCompare Component Should show 1`] = `
<div
  className="container"
>
  <div
    className="title"
  >
    title
  </div>
  <div
    className="graph"
  >
    <div
      className="lineContainer"
      key="0"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine first"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </div>
        <div
          className="graphAmount first"
        >
          8
        </div>
      </div>
    </div>
    <div
      className="lineContainer"
      key="1"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine"
            style={
              Object {
                "width": "20%",
              }
            }
          />
        </div>
        <div
          className="graphAmount"
        >
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GraphCompare Component Should show edge cases 1`] = `
<div
  className="container"
>
  <div
    className="title"
  >
    title
  </div>
  <div
    className="graph"
  >
    <div
      className="lineContainer"
      key="0"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine first"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
        <div
          className="graphAmount first"
        >
          0
          %
        </div>
      </div>
    </div>
    <div
      className="lineContainer"
      key="1"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine"
            style={
              Object {
                "width": "100%",
              }
            }
          />
        </div>
        <div
          className="graphAmount"
        >
          100
          %
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GraphCompare Component Should show percent 1`] = `
<div
  className="container"
>
  <div
    className="title"
  >
    title
  </div>
  <div
    className="graph"
  >
    <div
      className="lineContainer"
      key="0"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine first"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </div>
        <div
          className="graphAmount first"
        >
          80
          %
        </div>
      </div>
    </div>
    <div
      className="lineContainer"
      key="1"
    >
      <div>
        <div
          className="graphWrapper"
        >
          <div
            className="graphLine"
            style={
              Object {
                "width": "20%",
              }
            }
          />
        </div>
        <div
          className="graphAmount"
        >
          20
          %
        </div>
      </div>
    </div>
  </div>
</div>
`;
